### PR TITLE
[FIX] Change calculation of TR in glm.first_level.hemodynamic_models.compute_regressor()

### LINKED ISSUE #3360: glm.first_level.hemodynamic_models.compute_regressor() computes wrong TR
https://github.com/nilearn/nilearn/blob/9aa503f599f0835aea2bf9e2720bd5d3e33e941e/nilearn/glm/first_level/hemodynamic_models.py#L569

If your volumes are shifted by half a TR (like after fmriprep preprocessing) and your run is not very long, the TR is getting miscalculated/misestimated here (as is the case for us). Why not e.g. calculate the mean distance between the entries in `frame_times`?

### Nilearn version:
0.9.2.dev

### Expected behavior
return correct regressors

### Actual Behavior
cannot be sure if regressors are correct

Edit: Sorry, just saw the template. This is meant to be a bug report

### PR DESCRIPTION
**Pull request by askieslinger**
Closes #3360 .

Changes proposed in this pull request:

- change calculation of TR to be more precise 
   - use minimum distance between entries in `frame_times`
   - TR calculation not affected by possible shift in frame times (not starting at 0 but in the middle of first TR because of preprocessing)

### CODE REVIEW

**Review by NicolasGensollen — COMMENTED** [review]
Thanks for fixing this @askieslinger !
Could you add an entry in `doc/changes/latest.rst` under "bug fixes" as well as your name in `doc/changes/names.rst` as this is your first contribution to Nilearn?

**Comment by bthirion** [comment]
Thx ! 
Interestingly, this broke no test ...
Would you be so kind as to provide a test that enforces the new estimate ?

**Comment by askieslinger** [comment]
Yes, I'll look into doing this

> Thx ! Interestingly, this broke no test ... Would you be so kind as to provide a test that enforces the new estimate ?

**Comment by tsalo** [comment]
Here is a draft of something that could be the basis for a new test:

```python
import numpy as np

true_tr = 0.75
n_vols = 4
frame_times = np.linspace(true_tr / 2, (n_vols * true_tr) + (true_tr / 2), n_vols + 1)

# current version
estimated_tr = float(frame_times.max()) / (np.size(frame_times) - 1)
estimated_tr
# 0.84375

# proposed change
estimated_tr = float(np.min(frame_times[1:] - frame_times[:-1]))
estimated_tr
# 0.75
```

File: nilearn/glm/first_level/hemodynamic_models.py [hunk 1]
```diff
@@ -565,9 +565,8 @@ def compute_regressor(exp_condition, hrf_model, frame_times, con_id='cond',
         fir_delays = [int(x) for x in fir_delays]
     oversampling = int(oversampling)
 
-    # this is the average tr in this session, not necessarily the true tr
-    tr = float(frame_times.max()) / (np.size(frame_times) - 1)
-
+    # this is the minimal tr in this session, not necessarily the true tr
+    tr = float(np.min(frame_times[1:] - frame_times[:-1]))
```
Comment (tsalo) [inline nilearn/glm/first_level/hemodynamic_models.py:569]:
```suggestion
    tr = np.min(np.diff(frame_times))
```

Don't use this if you don't like it, but it's a bit simpler.

**Comment by askieslinger** [comment]
Thank you tsalo for your suggestions, will update this week.

**Comment by askieslinger** [comment]
> Here is a draft of something that could be the basis for a new test:
> 
> ```python
> import numpy as np
> 
> true_tr = 0.75
> n_vols = 4
> # create times for four volumes, shifted forward by half a TR (as with fMRIPrep slice timing corrected data)
> frame_times = np.linspace(true_tr / 2, (n_vols * true_tr) + (true_tr / 2), n_vols + 1)
> 
> # current version
> estimated_tr = float(frame_times.max()) / (np.size(frame_times) - 1)
> estimated_tr
> # 0.84375
> 
> # proposed change
> estimated_tr = float(np.min(frame_times[1:] - frame_times[:-1]))
> estimated_tr
> # 0.75
> ```

For this  direct test to work, I think we would need to extract estimating the TR into its own function, since the `compute_regressor` function does not return the TR? 

For an indirect test that is similar to the existing tests, we could use this, but it does not test the TR computation specifically. What would be the preferred solution?
````
true_tr = 0.75
n_vols = 4
hrf_model = 'spm'
# create times for four volumes, shifted forward by half a TR (as with fMRIPrep slice timing corrected data)
frame_times = np.linspace(true_tr / 2, (n_vols * true_tr) + (true_tr / 2), n_vols + 1)
reg,reg_names=compute_regressor(([0],[0],[100]),hrf_model, frame_times)
assert_almost_equal(np.sum(reg),0.4,1)
````

**Comment by bthirion** [comment]
I would suggest to isolate TR estimation in a function that would be called by the test. This would clarify the code.

**Comment by askieslinger** [comment]
Hi, so I added the test for the TR calculation. Unfortunately, I don't know how to resolve the merge conflicts without adding all the commits that happened in the meantime, sorry for that.

**Comment by Remi-Gau** [comment]
let me have a look

**Comment by Remi-Gau** [comment]
@askieslinger 

I used the github graphic interface to fix the conflicts: 
https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/addressing-merge-conflicts/resolving-a-merge-conflict-on-github

When you do that this will merge the main branch from the upstream repository into your branch into a single "merge commit".

Hope that helps.

**Review by ymzayek — APPROVED** [review]
LGTM thx @askieslinger

**Review by bthirion — APPROVED** [review]
LGTM, thx.

**Pull request merged**